mpi-proxy-split/mpi-wrappers/mpi_file_wrappers.cpp

```diff
@@ -42,19 +46,255 @@ USER_DEFINED_WRAPPER(int, File_open, (MPI_Comm) comm, (const char *) filename,
   JUMP_TO_LOWER_HALF(lh_info.fsaddr);
   retval = NEXT_FUNC(File_open)(realComm, filename, amode, info, fh);
   RETURN_TO_UPPER_HALF();
+
+  if (retval == MPI_SUCCESS) {
+    // Add file to virtual to real mapping
+    MPI_File virtFile = ADD_NEW_FILE(*fh);
+    *fh = virtFile;
+
+    // Save parameters
+    OpenFileParameters params;
+    params._mode = amode;
+    params._comm = comm;
+    params._info = info;
+    strcpy(params._filepath, filename);
+
+    // Mark offset/size/atomicity as uninitialized (so they won't be set) on
+    // restart unless they are manually set
+    params._datarep[0] = '\0';
+    params._offset = MPI_FILE_UNINITIALIZED;
+    params._size = MPI_FILE_UNINITIALIZED;
+    params._atomicity = MPI_FILE_UNINITIALIZED;
+    params._viewEType = MPI_FILE_UNINITIALIZED;
+    params._viewFType = MPI_FILE_UNINITIALIZED;
+    params._viewDisp = MPI_FILE_UNINITIALIZED;
+    params._viewInfo = MPI_FILE_UNINITIALIZED;
+
+    // Save this file handle to the global map
+    g_params_map[virtFile] = params;
+  }
+  DMTCP_PLUGIN_ENABLE_CKPT();
+  return retval;
+}
+
+USER_DEFINED_WRAPPER(int, File_get_atomicity, (MPI_File) fh, (int*) flag)
+{
+  int retval;
+  DMTCP_PLUGIN_DISABLE_CKPT();
+  MPI_File realFile = VIRTUAL_TO_REAL_FILE(fh);
+  JUMP_TO_LOWER_HALF(lh_info.fsaddr);
+  retval = NEXT_FUNC(File_get_atomicity)(realFile, flag);
+  RETURN_TO_UPPER_HALF();
+  DMTCP_PLUGIN_ENABLE_CKPT();
+  return retval;
+}
+
+USER_DEFINED_WRAPPER(int, File_set_atomicity, (MPI_File) fh, (int) flag)
+{
+  int retval;
+  DMTCP_PLUGIN_DISABLE_CKPT();
+  MPI_File realFile = VIRTUAL_TO_REAL_FILE(fh);
+  JUMP_TO_LOWER_HALF(lh_info.fsaddr);
+  retval = NEXT_FUNC(File_set_atomicity)(realFile, flag);
+  RETURN_TO_UPPER_HALF();
+  if (retval == MPI_SUCCESS) {
+    g_params_map[fh]._atomicity = flag;
+  }
+  DMTCP_PLUGIN_ENABLE_CKPT();
+  return retval;
+}
+
+USER_DEFINED_WRAPPER(int, File_set_size, (MPI_File) fh, (MPI_Offset) size)
+{
+  int retval;
+  DMTCP_PLUGIN_DISABLE_CKPT();
+  MPI_File realFile = VIRTUAL_TO_REAL_FILE(fh);
+  JUMP_TO_LOWER_HALF(lh_info.fsaddr);
+  retval = NEXT_FUNC(File_set_size)(realFile, size);
+  RETURN_TO_UPPER_HALF();
+  if (retval == MPI_SUCCESS) {
+    g_params_map[fh]._size = size;
+  }
   DMTCP_PLUGIN_ENABLE_CKPT();
   return retval;
 }
 
-DEFINE_FNC(int, File_close, (MPI_File *) fh);
-DEFINE_FNC(int, File_get_size, (MPI_File) fh, (MPI_Offset *) size);
-DEFINE_FNC(int, File_read_at, (MPI_File) fh, (MPI_Offset) offset, (void *) buf,
-           (int) count, (MPI_Datatype) datatype, (MPI_Status *) status);
+USER_DEFINED_WRAPPER(int, File_get_size, (MPI_File) fh, (MPI_Offset *) size)
+{
+  int retval;
+  DMTCP_PLUGIN_DISABLE_CKPT();
+  MPI_File realFile = VIRTUAL_TO_REAL_FILE(fh);
+  JUMP_TO_LOWER_HALF(lh_info.fsaddr);
+  retval = NEXT_FUNC(File_get_size)(realFile, size);
```

Comment:
Since you are saving the size every time `MPI_FIle_set_size` in `g_params_map[fh]._size`, do you need to do an actual call or can you just return the information from the global map?
In other words, are there other APIs that can change the size than `MPI_File_set_size` API?

Reply:
It appears other APIs can alter the size as well (such as writing to the file). I will change it to be treated the same as the file offset.

Reply:
Actually, this means it doesn't have to be saved and restored at all.